iree/compiler/Dialect/Flow/Transforms/DispatchIDCanonicalizations.cpp

```diff
@@ -0,0 +1,353 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#define DEBUG_TYPE "iree-flow-canonicalize-dispatch-ids"
+
+/// This file contains patterns allowing canonicalization taking advantage  of
+/// flow dispatch ops. When the ID range is know we may be able to calculate the
+/// min/max of the derived values and may allow folding affine.min or ForOp.
+/// This file duplicate some of the logic from
+/// `linalg::AffineMinSCFCanonicalizationPattern`. In the future we could unify
+/// this code back by upstreaminng an interface to MLIR to represent generic ID
+/// ops.
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+/// Traverse `e` and return an AffineExpr where all occurrences of `dim` have
+/// been replaced by either:
+///  - `min` if `positivePath` is true when we reach an occurrence of `dim`
+///  - `max` if `positivePath` is true when we reach an occurrence of `dim`
+/// `positivePath` is negated each time we hit a multiplicative or divisive
+/// binary op with a constant negative coefficient.
+static AffineExpr substWithMin(AffineExpr e, AffineExpr dim, AffineExpr min,
+                               AffineExpr max, bool positivePath = true) {
+  if (e == dim) return positivePath ? min : max;
```

Comment:
Maybe a bit more readable if there is a `getMinOfExpr` and `getMaxOfExpr`. Those are useful utilities to  have in general I think. I have encountered this a couple of times and was going to add this upstream (but I in the end didnt really needed it so didnt go through with it)

Reply:
Actually maybe this comment is wrong. I see how `positivePath` makes this more involved. But nevertheless leaving the comment here in case this make sense to you